### ISSUE TITLE
Fine-tuning T5 models with multiprocessing_train

### ISSUE THREAD

**Issue by lyy1994** [facebookresearch/ParlAI#4430]
**Bug description**
I am trying to fine-tune a T5-small model via `multiprocessing_train`. I am using ParlAI 1.5.1 release and 4 V100 with 16GB RAM. However, it seems that the process hangs and multiple unexpected processes spawn in the first device.

**Reproduction steps**
The command I used is simply modified from the [docs](https://parl.ai/docs/agent_refs/hugging_face.html):

```
parlai multiprocessing_train -m hugging_face/t5 -mf /tmp/model_file -t convai2 -bs 24 --fp16 true -eps 1 -lr 1e-5 --optimizer adam --t5-model-arch t5-small
```

However, everything works fine if I use `train_model`:

```
parlai train_model -m hugging_face/t5 -mf /tmp/model_file -t convai2 -bs 24 --fp16 true -eps 1 -lr 1e-5 --optimizer adam --t5-model-arch t5-small
```

**Expected behavior**
It is expected that the program should produce only 4 processes, each for one device.

**Logs**

However, as you may see in the figure below, 3 unexpected processes appear in the first device.
![image](https://user-images.githubusercontent.com/7354097/159018982-3addca68-de9a-4b4d-9342-aa0e5222625e.png)

And no new log was printed afterwards.
```
22:02:47 | Distributed group initialized
22:02:53 | building dictionary first...
22:02:53 | No model with opt yet at: /tmp/model_file(.opt)
22:02:53 | Using CUDA
22:03:06 | Total parameters: 60,506,624 (60,506,624 trainable)
22:03:06 | Opt:
22:03:06 |     adafactor_eps: '(1e-30, 0.001)'
22:03:06 |     adam_eps: 1e-08
22:03:06 |     add_p1_after_newln: False
22:03:06 |     aggregate_micro: False
22:03:06 |     allow_missing_init_opts: False
22:03:06 |     batchsize: 24
22:03:06 |     beam_block_full_context: True
22:03:06 |     beam_block_list_filename: None
22:03:06 |     beam_block_ngram: -1
22:03:06 |     beam_context_block_ngram: -1
22:03:06 |     beam_delay: 30
22:03:06 |     beam_length_penalty: 0.65
22:03:06 |     beam_min_length: 1
22:03:06 |     beam_size: 1
22:03:06 |     betas: '(0.9, 0.999)'
22:03:06 |     bpe_add_prefix_space: None
22:03:06 |     bpe_debug: False
22:03:06 |     bpe_dropout: None
22:03:06 |     bpe_merge: None
22:03:06 |     bpe_vocab: None
22:03:06 |     compute_tokenized_bleu: False
22:03:06 |     datapath: /data/liyanyang/ParlAI/data
22:03:06 |     datatype: train
22:03:06 |     ddp_backend: ddp
22:03:06 |     delimiter: '\n'
22:03:06 |     dict_class: parlai.core.dict:DictionaryAgent
22:03:06 |     dict_endtoken: __end__
22:03:06 |     dict_file: /tmp/model_file.dict
22:03:06 |     dict_include_test: False
22:03:06 |     dict_include_valid: False
22:03:06 |     dict_initpath: None
22:03:06 |     dict_language: english
22:03:06 |     dict_lower: False
22:03:06 |     dict_max_ngram_size: -1
22:03:06 |     dict_maxexs: -1
22:03:06 |     dict_maxtokens: -1
22:03:06 |     dict_minfreq: 0
22:03:06 |     dict_nulltoken: __null__
22:03:06 |     dict_starttoken: __start__
22:03:06 |     dict_textfields: text,labels
22:03:06 |     dict_tokenizer: re
22:03:06 |     dict_unktoken: __unk__
22:03:06 |     display_examples: False
22:03:06 |     distributed_world_size: 4
22:03:06 |     download_path: None
22:03:06 |     dynamic_batching: None
22:03:06 |     embedding_projection: random
22:03:06 |     embedding_type: random
22:03:06 |     eval_batchsize: None
22:03:06 |     eval_dynamic_batching: None
22:03:06 |     evaltask: None
22:03:06 |     final_extra_opt:
22:03:06 |     force_fp16_tokens: False
22:03:06 |     fp16: True
22:03:06 |     fp16_impl: safe
22:03:06 |     gpu: 0
22:03:06 |     gradient_clip: 0.1
22:03:06 |     hide_labels: False
22:03:06 |     history_add_global_end_token: None
22:03:06 |     history_reversed: False
22:03:06 |     history_size: -1
22:03:06 |     image_cropsize: 224
22:03:06 |     image_mode: raw
22:03:06 |     image_size: 256
22:03:06 |     inference: greedy
22:03:06 |     init_model: None
22:03:06 |     init_opt: None
22:03:06 |     interactive_mode: False
22:03:06 |     invsqrt_lr_decay_gamma: -1
22:03:06 |     is_debug: False
22:03:06 |     label_truncate: None
22:03:06 |     learningrate: 1e-05
22:03:06 |     load_from_checkpoint: True
22:03:06 |     log_every_n_secs: -1
22:03:06 |     log_every_n_steps: 50
22:03:06 |     loglevel: info
22:03:06 |     lr_scheduler: reduceonplateau
22:03:06 |     lr_scheduler_decay: 0.5
22:03:06 |     lr_scheduler_patience: 3
22:03:06 |     max_train_steps: -1
22:03:06 |     max_train_time: -1
22:03:06 |     metrics: default
22:03:06 |     model: hugging_face/t5
22:03:06 |     model_file: /tmp/model_file
22:03:06 |     momentum: 0
22:03:06 |     multiprocessing: True
22:03:06 |     multitask_weights: [1]
22:03:06 |     mutators: None
22:03:06 |     nesterov: True
22:03:06 |     no_cuda: False
22:03:06 |     num_epochs: 1.0
22:03:06 |     num_workers: 0
22:03:06 |     nus: (0.7,)
22:03:06 |     optimizer: adam
22:03:06 |     override: "{'fp16': True, 'optimizer': 'adam', 't5_model_arch': 't5-small', 'gpu': 0}"
22:03:06 |     parlai_home: /data/liyanyang/ParlAI
22:03:06 |     person_tokens: False
22:03:06 |     port: None
22:03:06 |     rank: 0
22:03:06 |     rank_candidates: False
22:03:06 |     save_after_valid: False
22:03:06 |     save_every_n_secs: -1
22:03:06 |     short_final_eval: False
22:03:06 |     skip_generation: False
22:03:06 |     special_tok_lst: None
22:03:06 |     split_lines: False
22:03:06 |     starttime: Mar18_22-02
22:03:06 |     t5_dropout: 0.0
22:03:06 |     t5_generation_config: None
22:03:06 |     t5_model_arch: t5-small
22:03:06 |     t5_model_parallel: False
22:03:06 |     task: convai2
22:03:06 |     temperature: 1.0
22:03:06 |     tensorboard_log: False
22:03:06 |     tensorboard_logdir: None
22:03:06 |     text_truncate: None
22:03:06 |     topk: 10
22:03:06 |     topp: 0.9
22:03:06 |     truncate: -1
22:03:06 |     update_freq: 1
22:03:06 |     use_reply: label
22:03:06 |     validation_cutoff: 1.0
22:03:06 |     validation_every_n_epochs: -1
22:03:06 |     validation_every_n_secs: -1
22:03:06 |     validation_every_n_steps: -1
22:03:06 |     validation_max_exs: -1
22:03:06 |     validation_metric: accuracy
22:03:06 |     validation_metric_mode: None
22:03:06 |     validation_patience: 10
22:03:06 |     validation_share_agent: False
22:03:06 |     verbose: False
22:03:06 |     wandb_entity: None
22:03:06 |     wandb_log: False
22:03:06 |     wandb_name: None
22:03:06 |     wandb_project: None
22:03:06 |     warmup_rate: 0.0001
22:03:06 |     warmup_updates: -1
22:03:06 |     weight_decay: None
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
22:03:07 | Current ParlAI commit: df97e1fa3dab5b5e2050a247c04601f0d43e4d28
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
22:03:07 | creating task(s): convai2
22:03:07 | loading fbdialog data: /data/liyanyang/ParlAI/data/ConvAI2/train_self_original.txt
22:03:11 | Using --num-epochs in distributed mode can lead to slowdowns. See https://github.com/facebookresearch/ParlAI/pull/3379 for more info.
22:03:11 | training...
```

**Additional context**
`huggingface` also gives a warning as shown in the log: `huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...`. Does it matter? I have tried to attach `TOKENIZERS_PARALLELISM=false` before the command and it disables the warning but does not help in the above problem.

**Comment by stephenroller** [comment]
Hm that looks like it's using model parallelism too. Can you try setting `--t5-model-parallel false --model-parallel false`?

**Comment by lyy1994** [comment]
Thanks for your reply. I tried to add `--t5-model-parallel false` (adding `--model-parallel false` will raise `Parse Error: unrecognized arguments`). However, the same thing still happens:

![image](https://user-images.githubusercontent.com/7354097/159102481-d393fb40-465d-452d-a518-0de311f4a2b3.png)

**Comment by stephenroller** [comment]
@klshuster any idea why t5 is model paralleling?

(As a mitigation, you could just use regular train with a 4x batch size.)

**Comment by lyy1994** [comment]
I am using `train_model` now. But it is slow...

**Comment by klshuster** [comment]
Thanks for flagging, fix is up in #4434

**Comment by lyy1994** [comment]
Thanks! It works now!
